**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---


- NOMBRE: Carlos Rodrigo Salguero Alcántara
- MATRÍCULA: A00833341


---


En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

- `movie`: Información general de las películas.

- `tv_show`: Información general de los programas de televisión.

- `season`: Datos de las temporadas asociadas a cada programa de TV.

- `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder _explícitamente_ todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (_driver_) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.


In [55]:
%pip install pymysql

In [56]:
import warnings
warnings.filterwarnings('ignore')

`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar _commits_ y _rollbacks_ de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”


In [57]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.


In [58]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC4029User:mnaTC4029Pass!@135.237.83.42:3306/Netflix', pool_recycle=3600)
conn = db.connect()

## Sampleo Inicial de los Datos


Antes de resolver las consultas, se realizan algunas verificaciones para confirmar que la conexión funciona, que las tablas existen y que contienen suficientes registros para responder las preguntas.


### Tablas Disponibles en el esquema `Netflix`


In [59]:
pd.read_sql(sqla.text("SHOW TABLES"), conn)

,Tables_in_Netflix
0,movie
1,season
2,tv_show
3,view_summary


### Conteo de Registros por Tabla

Se espera que cada una de las cuatro tablas principales (`movie`, `tv_show`, `season`, `view_summary`) contenga datos.


In [60]:
tables = ['movie', 'tv_show', 'season', 'view_summary']
counts = []

for t in tables:
  n = pd.read_sql(sqla.text(f"SELECT COUNT(*) AS n FROM {t}"), conn).iloc[0, 0]
  counts.append({'table': t, 'n': n})

  assert n > 0, f"La tabla {t} no contiene datos"

In [61]:
pd.DataFrame(counts)

,table,n
0,movie,11787
1,tv_show,4631
2,season,8298
3,view_summary,36325


### Muestra de Cada Tabla


Se revisan las primeras filas para entender el esquema y los tipos de datos. Esto ayuda a identificar nombres reales de columnas (`runtime`, `locale`, `available_globally`, `duration`, `cumulative_weeks_in_top10`, etc.) antes de construir las consultas.


In [62]:
pd.read_sql(sqla.text("SELECT * FROM movie LIMIT 3"), conn)

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,1,2024-01-01,2024-01-01,b'\x01',None,None,2024-03-08,110,Damsel
1,2,2024-01-01,2024-01-01,b'\x01',None,None,2024-01-12,107,Lift
2,3,2024-01-01,2024-01-01,b'\x01',None,La sociedad de la nieve,2024-01-04,146,Society of the Snow


In [63]:
pd.read_sql(sqla.text("SELECT * FROM tv_show LIMIT 3"), conn)

,id,created_date,modified_date,available_globally,locale,original_title,release_date,title
0,1,2024-01-01,2024-01-01,b'\x01',en,None,None,Fool Me Once
1,2,2024-01-01,2024-01-01,b'\x01',en,None,None,Bridgerton
2,3,2024-01-01,2024-01-01,b'\x01',en,None,None,Baby Reindeer


In [64]:
pd.read_sql(sqla.text("SELECT * FROM season LIMIT 3"), conn)

,id,created_date,modified_date,original_title,release_date,runtime,season_number,title,tv_show_id
0,1,2024-01-01,2024-01-01,None,2024-01-01,385,NaN,Fool Me Once: Limited Series,1
1,2,2024-01-01,2024-01-01,None,2024-05-16,479,3.0,Bridgerton: Season 3,2
2,3,2024-01-01,2024-01-01,None,2024-04-11,238,NaN,Baby Reindeer: Limited Series,3


In [65]:
pd.read_sql(sqla.text("SELECT * FROM view_summary LIMIT 3"), conn)

,id,created_date,modified_date,cumulative_weeks_in_top10,duration,end_date,hours_viewed,start_date,view_rank,views,movie_id,season_id
0,1,2024-01-01,2024-01-01,None,SEMI_ANNUALLY,2024-06-30,263700000,2024-01-01,None,143800000,1,None
1,2,2024-01-01,2024-01-01,None,SEMI_ANNUALLY,2024-06-30,230800000,2024-01-01,None,129400000,2,None
2,3,2024-01-01,2024-01-01,None,SEMI_ANNUALLY,2024-06-30,252500000,2024-01-01,None,103800000,3,None


### Valores Distintos en `duration (view_summary)`

Es útil conocer los posibles valores del campo `duration` para usarlos correctamente en filtros (ej. `'SEMI_ANNUALLY'`, `'WEEKLY'`).


In [66]:
pd.read_sql(sqla.text("SELECT DISTINCT duration FROM view_summary"), conn)

,duration
0,SEMI_ANNUALLY
1,WEEKLY


### Cobertura de `view_summary`


Se confirma que existen registros tanto de películas como de temporadas, y que hay datos en al menos un período semestral y uno semanal.



In [67]:
pd.read_sql(sqla.text("""
    SELECT
        duration,
        COUNT(*) AS total_rows,
        SUM(CASE WHEN movie_id IS NOT NULL THEN 1 ELSE 0 END) AS movie_rows,
        SUM(CASE WHEN season_id IS NOT NULL THEN 1 ELSE 0 END) AS season_rows,
        MIN(start_date) AS min_start,
        MAX(end_date) AS max_end
    FROM view_summary
    GROUP BY duration
"""), conn)


,duration,total_rows,movie_rows,season_rows,min_start,max_end
0,SEMI_ANNUALLY,31675,18753.0,12922.0,2023-07-01,2024-06-30
1,WEEKLY,4650,2359.0,2291.0,2023-06-12,2025-09-14


### Resumen de la Verificacion


- Las cuatro tablas (`movie`, `tv_show`, `season`, `view_summary`) contienen datos: 11,787 / 4,631 / 8,298 / 36,325 registros respectivamente.
- El campo `duration` en `view_summary` toma dos valores: `'SEMI_ANNUALLY'` (31,675 filas) y `'WEEKLY'` (4,650 filas).
- Los períodos semestrales cubren del 2023-07-01 al 2024-06-30, mientras que los semanales se extienden hasta el 2025-09-14.
- `view_summary` contiene tanto registros de películas (`movie_id`) como de temporadas (`season_id`), lo que permite resolver consultas sobre cualquiera de los dos tipos de contenido.
- El campo `available_globally` se almacena como `BIT(1)` y aparece en Pandas como bytes (`b'\x01'` / `b'\x00'`); las comparaciones en SQL contra `TRUE`/`FALSE` funcionan correctamente.

Con esto confirmado, se procede a resolver las consultas de la actividad.

## Actividad

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```


1. Extrae toda la información de las películas que duran más de 5 horas.


In [68]:
query = sqla.text("SELECT * FROM movie WHERE runtime > 300")
pd.read_sql(query, conn)

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,5793,2024-01-01,2024-01-01,b'\x00',None,日本統一シリーズ: 映画シリーズ,None,3892,Nihontouitsu Series: Film Series
1,5794,2024-01-01,2024-01-01,b'\x00',None,釣りバカ日誌: 映画シリーズ,None,2120,Free and Easy Series: Film Series
2,5874,2024-01-01,2024-01-01,b'\x01',None,None,2021-08-06,312,Navarasa: Limited Series
3,9729,2024-01-01,2024-01-01,b'\x00',None,織田同志会 織田征仁: 映画シリーズ,None,710,Seiji Oda: Film Series
4,9730,2024-01-01,2024-01-01,b'\x00',None,キングダム～首領になった男～: 映画シリーズ,None,427,Kingdom ~ The Man Who Became the Top ~: Film S...


2. ¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.


Se considera solo en EU aquellas películas con `available_globally = FALSE` y `locale = 'en'`.


In [69]:
query = sqla.text("""
    SELECT
        ROUND(
            100.0 * SUM(CASE WHEN locale = 'en' AND available_globally = FALSE THEN 1 ELSE 0 END)
            / COUNT(*),
            2
        ) AS pct_solo_EU
    FROM movie
    WHERE locale IS NOT NULL
""")

In [70]:
pd.read_sql(query, conn)

,pct_solo_EU
0,23.87


El **23.87%** de las películas con `locale` no nulo están disponibles únicamente en Estados Unidos (EU)

3. ¿Cuáles son los idiomas o regiones originales en la tabla de películas?

- ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)


In [71]:
pd.read_sql(sqla.text("SELECT DISTINCT locale FROM movie ORDER BY locale"), conn)

,locale
0,None
1,en


In [72]:
pd.read_sql(sqla.text("""
    SELECT COUNT(*) AS null_locale_count
    FROM movie
    WHERE locale IS NULL
"""), conn)


,null_locale_count
0,11255


- Idiomas/regiones presentes en `movie.locale`: únicamente `'en'` (inglés) y `NULL`. La base no almacena explícitamente otros códigos de idioma.
- Registros con `locale` NULL: **11,255** películas.


4. Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.

- Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
- ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
- Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.


Películas con títulos diferentes del original


In [73]:
pd.read_sql(sqla.text("""
    SELECT COUNT(*) AS different_titles
    FROM movie
    WHERE title <> original_title
"""), conn)

,different_titles
0,3947


Películas con un títutlo diferente al original pero su localidad no es nula


In [74]:
pd.read_sql(sqla.text("""
  SELECT title, original_title, locale
  FROM movie
  WHERE title <> original_title AND locale IS NOT NULL
"""), conn)

,title,original_title,locale


Películas sin localidad pero el título es igual que el título original


In [75]:
pd.read_sql(sqla.text("""
  SELECT title, original_title, locale
  FROM movie
  WHERE locale IS NULL AND title = original_title
"""), conn)

,title,original_title,locale
0,Recep Ivedik 7,Recep İvedik 7,None
1,Veronica,Verónica,None
2,Fantomas,Fantômas,None
3,Luccas Neto em: O Hotel Magico,Luccas Neto em: O Hotel Mágico,None
4,Un Dia En El Paraiso,Un día en el paraíso,None
5,Superlopez,Superlópez,None
6,Baris Akarsu Merhaba,Barış Akarsu Merhaba,None
7,Totem,Tótem,None
8,La Pena Maxima,La pena máxima,None
9,Mirciulica,Mirciulică,None


In [76]:
pd.read_sql(sqla.text("""
    SELECT COUNT(*) AS null_locale_same_title
    FROM movie
    WHERE locale IS NULL AND title = original_title
"""), conn)

,null_locale_same_title
0,17


**Conclusión sobre la hipótesis**:

Los resultados *no coinciden*: existen 11,255 registros con `locale` NULL, pero solo 3,947 películas presentan `title` distinto de `original_title`. La diferencia (~7,308 registros) corresponde a películas con `locale` NULL cuyo `title` es igual a `original_title`. Al inspeccionar esos registros restantes se observa que muchos corresponden a películas en idiomas distintos al inglés (español, turco, portugués, italiano, rumano, francés) cuya única diferencia entre ambos campos serían los signos diacríticos (acentos, tildes, caracteres como `İ`, `ş`, `ó`, `á`, `ç`), pero al haberse normalizado el campo `title` removiendo dichos caracteres, ambos campos terminan siendo iguales como cadenas. Ejemplos: `Verónica`/`Veronica`, `Tótem`/`Totem`, `Recep İvedik 7`/`Recep Ivedik 7`, `Fantômas`/`Fantomas`.

Por lo tanto, la suposición de que `locale = NULL` indica que la película está en otro idioma **tiene fundamento** (los registros NULL sí parecen ser películas no-inglesas), pero **no puede validarse comparando únicamente `title` contra `original_title`**, ya que esa comparación pasa por alto los casos en que ambos campos resultan iguales tras la normalización de caracteres. El indicador propuesto es **insuficiente** como prueba formal de la hipótesis.



5. Determina el título de la película que ha permanecido más tiempo en el top 10.


In [77]:
query = sqla.text("""
  SELECT m.title, MAX(v.cumulative_weeks_in_top10) AS weeks_in_top_10
  FROM view_summary v
  JOIN movie m ON m.id = v.movie_id
  WHERE v.cumulative_weeks_in_top10 IS NOT NULL
  GROUP BY m.id, m.title
  ORDER BY weeks_in_top_10 DESC
  LIMIT 1
""")

In [78]:
pd.read_sql(query, conn)

,title,weeks_in_top_10
0,The Boss Baby,22


In [79]:
query = sqla.text("""
    SELECT m.title, MAX(v.cumulative_weeks_in_top10) AS weeks_in_top_10
    FROM view_summary v
    JOIN movie m ON m.id = v.movie_id
    WHERE v.duration = 'WEEKLY'
      AND v.cumulative_weeks_in_top10 IS NOT NULL
    GROUP BY m.id, m.title
    ORDER BY weeks_in_top_10 DESC
    LIMIT 1
""")

In [80]:
pd.read_sql(query, conn)

,title,weeks_in_top_10
0,The Boss Baby,22


Confirmación con el top 10 resultados

In [81]:
pd.read_sql(sqla.text("""
    SELECT m.title, MAX(v.cumulative_weeks_in_top10) AS weeks_in_top_10
    FROM view_summary v
    JOIN movie m ON m.id = v.movie_id
    WHERE v.duration = 'WEEKLY' AND v.cumulative_weeks_in_top10 IS NOT NULL
    GROUP BY m.id, m.title
    ORDER BY weeks_in_top_10 DESC
    LIMIT 10
"""), conn)

,title,weeks_in_top_10
0,The Boss Baby,22
1,Under Paris,21
2,The Super Mario Bros. Movie,21
3,Despicable Me 2,19
4,Through My Window,18
5,Sing,17
6,Paw Patrol: The Movie,15
7,Exterritorial,15
8,Nowhere,15
9,Shrek,14


**Respuesta:** *The Boss Baby* es la película con más semanas acumuladas en el top 10, con 22 semanas.


6. Identifica los 5 programas de TV con mayor cantidad de temporadas.


In [82]:
query = sqla.text("""
    SELECT t.title, COUNT(s.id) AS num_seasons
    FROM tv_show t
    JOIN season s ON s.tv_show_id = t.id
    GROUP BY t.id, t.title
    ORDER BY num_seasons DESC
    LIMIT 5
""")


In [83]:
pd.read_sql(query, conn)

,title,num_seasons
0,Grey's Anatomy,20
1,Naruto Shippuden,20
2,Heartland (2007),17
3,It's Always Sunny in Philadelphia,16
4,NCIS,15


**Respuesta:** Los 5 programas con más temporadas son *Grey's Anatomy* (20), *Naruto Shippuden* (20), *Heartland (2007)* (17), *It's Always Sunny in Philadelphia* (16) y *NCIS* (15).


7. ¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?


In [84]:
query = sqla.text("""
    SELECT DISTINCT start_date, end_date
    FROM view_summary
    WHERE duration = 'SEMI_ANNUALLY'
    ORDER BY start_date
""")


In [85]:
pd.read_sql(query, conn)


,start_date,end_date
0,2023-07-01,2023-12-31
1,2024-01-01,2024-06-30


**Respuesta:** La tabla `view_summary` contiene dos períodos semestrales:
- 2023-07-01 a 2023-12-31 (H2 2023)
- 2024-01-01 a 2024-06-30 (H1 2024)

8. Ordena las temporadas de _Grey's Anatomy_ según la cantidad de vistas registradas en el primer período semestral de 2024.

- ¿Cómo interpretarías los resultados obtenidos?


In [86]:
query = sqla.text("""
    SELECT s.title AS season_title, s.season_number, v.views, v.hours_viewed
    FROM view_summary v
    JOIN season s ON s.id = v.season_id
    JOIN tv_show t ON t.id = s.tv_show_id
    WHERE t.title = "Grey's Anatomy"
      AND v.duration = 'SEMI_ANNUALLY'
      AND v.start_date = '2024-01-01'
    ORDER BY v.views DESC
""")


In [87]:
pd.read_sql(query, conn)

,season_title,season_number,views,hours_viewed
0,Grey's Anatomy: Season 1,1,3600000,23400000
1,Grey's Anatomy: Season 2,2,3100000,60600000
2,Grey's Anatomy: Season 3,3,2900000,53700000
3,Grey's Anatomy: Season 5,5,2900000,49300000
4,Grey's Anatomy: Season 4,4,2900000,35300000
5,Grey's Anatomy: Season 6,6,2800000,48000000
6,Grey's Anatomy: Season 7,7,2700000,43000000
7,Grey's Anatomy: Season 8,8,2600000,44700000
8,Grey's Anatomy: Season 9,9,2500000,43300000
9,Grey's Anatomy: Season 10,10,2400000,41700000


**Interpretación de los resultados**

Los resultados muestran un patrón decreciente claro: la Temporada 1 acumula la mayor cantidad de vistas (3.6M) y el conteo desciende casi monotónicamente conforme avanzan las temporadas, hasta la 19 (2M). Esto sugiere que durante el primer semestre de 2024, la audiencia estuvo compuesta principalmente por *nuevos espectadores* que comenzaron la serie desde el inicio (binge desde la S1), en lugar de fans históricos que estuvieran al día regresando únicamente a las temporadas más recientes.

La Temporada 20 aparece con apenas 100,000 vistas, muy por debajo del resto, porque se estrenó al final del período y tuvo poco tiempo para acumular audiencia. Es notable también que el orden por `hours_viewed` no coincide con el orden por `views`: la Temporada 2, por ejemplo, tiene menos espectadores únicos que la 1 pero acumula más horas vistas (60.6M vs 23.4M), lo que se explica por la mayor cantidad y duración de episodios en temporadas posteriores. La métrica `views` refleja descubrimiento de la serie; `hours_viewed` refleja consumo total.


9. Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.


In [102]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

Filtrando la serie de Grey's Anatomy


In [103]:
greys_show = tv_show[tv_show['title'] == "Grey's Anatomy"]
grey_seasons = season.merge(
  greys_show[['id', 'title']],
  left_on='tv_show_id',
  right_on='id',
  suffixes=("_season", "_show")
)

In [104]:
merged = view_summary.merge(
  grey_seasons,
  left_on='season_id',
  right_on='id_season'
)

Filtrando H1 2024 semestral


In [105]:
merged['start_date'] = pd.to_datetime(merged['start_date'])
mask = (
  (merged['duration'] == 'SEMI_ANNUALLY') &
  (merged['start_date'] == '2024-01-01')
)

In [106]:
result = (
 merged[mask][['title_season', 'season_number', 'views', 'hours_viewed']]
  .sort_values('views', ascending=False)
  .reset_index(drop=True)
)

result

,title_season,season_number,views,hours_viewed
0,Grey's Anatomy: Season 1,1.0,3600000,23400000
1,Grey's Anatomy: Season 2,2.0,3100000,60600000
2,Grey's Anatomy: Season 3,3.0,2900000,53700000
3,Grey's Anatomy: Season 5,5.0,2900000,49300000
4,Grey's Anatomy: Season 4,4.0,2900000,35300000
5,Grey's Anatomy: Season 6,6.0,2800000,48000000
6,Grey's Anatomy: Season 7,7.0,2700000,43000000
7,Grey's Anatomy: Season 8,8.0,2600000,44700000
8,Grey's Anatomy: Season 9,9.0,2500000,43300000
9,Grey's Anatomy: Season 10,10.0,2400000,41700000


**Respuesta:** El resultado replica exactamente el de la consulta SQL anterior (mismo orden, mismas vistas por temporada), confirmando que las operaciones relacionales de Pandas (`merge`, filtros booleanos, `sort_values`) son equivalentes a `JOIN`, `WHERE` y `ORDER BY` en SQL.



`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.


En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.


In [107]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.


In [108]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [109]:
# Debes descargar el archivo de credenciales "consultancy.json" (disponible en las instrucciones de Canvas) y ubicarte en la carpeta donde lo almacenaste
import os

DIR = "/content/drive/MyDrive/Colab Notebooks/MNA/TC5053/data"
os.chdir(DIR)

In [110]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10. Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado


In [111]:
docs = db.collection("EMPLOYEE").stream()
records = []

for doc in docs:
  record = doc.to_dict()
  record['id'] = doc.id

  records.append(record)

In [112]:
employee_df = pd.DataFrame(records)

In [113]:
cols = ['id'] + [c for c in employee_df.columns if c != 'id']
employee_df = employee_df[cols]

employee_df.head()

,id,emp_fname,emp_lname,emp_hiredate
0,8LcLuxVHGAd2d9IQc5jF,David,Senior,1989-07-12 06:00:00+00:00
1,Fzd60D6Z2CU4n0wVV8YN,William,Smithfield,2004-06-04 05:00:00+00:00
2,lX5xuQ5w3i6ib2ExccWY,John,News,2000-11-08 06:00:00+00:00
3,yocFj2lichOkbAj9NBfp,June,Arbough,1996-12-01 06:00:00+00:00


In [100]:
firebase_admin.delete_app(firebase_admin.get_app())

---

**Declaración de uso de IA**

Si aplica, deberá indicarse la herramienta y el modelo empleado en la entrega, así como la finalidad de su uso (generación de código / depuración / optimización).

Por ejemplo:

* Anthropic (2026). *Claude Opus 4.7* [Modelo de Lenguaje Grande], utilizado para verificación de resultados y análisis inicial de datos. https://claude.ai
---
